lua/entities/gmod_wire_expression2/base/optimizer.lua

```diff
@@ -0,0 +1,182 @@
+--[[
+An optimizer for E2 abstract syntax trees, as produced by the parser and
+consumed by the compiler.
+
+Currently it only performs some simple peephole optimizations and constant
+propagation. Ideally, we'd do type inference as much as possible before
+optimizing, which would give us more useful information throughout.
+--]]
+
+E2Lib.Optimizer = {}
+local Optimizer = E2Lib.Optimizer
+Optimizer.__index = Optimizer
+
+local optimizerDebug = CreateConVar("wire_expression2_optimizer_debug", 0,
```

Comment:
e2 already has a debug convar, why not reuse that?

Reply:
depends, how chatty is this?

Reply:
This outputs a potentially huge amount of text (the AST of the optimized program) on every compilation.